### PR TITLE
Respect .editorconfig `end_of_line` and `charset` in generated code

### PR DESCRIPTION
**Pull request by YoelDruxman**
# Respect .editorconfig `end_of_line` and `charset` in generated code

## Description

Generated mapper files (`.g.cs`) now respect `.editorconfig` settings for line endings and character encoding.

### Problem

Generated code always used CRLF (`\r\n`) line endings and UTF-8 without BOM, regardless of the project's `.editorconfig` configuration. This caused issues for projects configured with different line endings.

### Solution

Read `end_of_line` and `charset` settings from the source file's `.editorconfig` via `AnalyzerConfigOptionsProvider` and apply them to the generated output.

**Supported settings:**
- `end_of_line`: `lf`, `crlf`, `cr` (defaults to `crlf`)
- `charset`: `utf-8`, `utf-8-bom`, `utf-16be`, `utf-16le` (defaults to `utf-8` without BOM)

### Known Limitation
Due to a Roslyn limitation, editorconfig settings are read from the **source file** containing the mapper declaration, not from generated `.g.cs` files. This means `[*.g.cs]` patterns in `.editorconfig` will not be applied. Users should configure settings in `[*.cs]` or `[*]` sections.

**Files changed:**
- `LineEndingTextWriter.cs` - Streaming line ending replacement
- `IncrementalValuesProviderExtensions.cs` - Read editorconfig settings and apply encoding
- `MapperNode.cs` - Store `EndOfLine` and `Charset` properties
- `MapperGenerator.cs` - Pass settings from source file to output

Addresses [this discussion](https://github.com/riok/mapperly/discussions/2094)

## Checklist
  - [x] The existing code style is followed
  - [x] The commit message follows our guidelines
  - [ ] Performed a self-review of my code
  - [x] Hard-to-understand areas of my code are commented
  - [x] The documentation is updated (as applicable)
  - [x] Unit tests are added/updated
  - [x] Integration tests are added/updated (as applicable)

### CODE REVIEW

**Comment by latonz** [comment]
Thank you for this contribution. Before I start the review, I’d like to clarify a few open questions:
* How do other source generators handle this? It seems like a challenge all source generators face.
* What is the performance impact?
* Would it make more sense to emit the correct line feeds directly in the syntax emitters instead of replacing them afterward? This would probably reduce the performance impact.

**Comment by YoelDruxman** [comment]
Thank you @latonz! I am still working on this. Let me consider your feedback and update the PR or at least report back.

**Comment by YoelDruxman** [comment]
@latonz I think this is ready for your review. Your thoughtful questions led to changing the implementation for the better:

1. It seems that this is an unsolved problem.
2. The performance has been improved to what I believe is almost transparent.
3. This seems like more effort and difficulty than it is worth.

[Here is a report](https://github.com/user-attachments/files/24652667/EDITORCONFIG_RESEARCH.md) I created with Claude that delves a little deeper.

Please let me know what you think. Any feedback is welcome! Thank you for this wonderful library!

File: docs/docs/configuration/generated-source.mdx [hunk 1]
```diff
@@ -84,3 +84,44 @@ The code generated by Mapperly can be identified in several ways:
   > Note: the version `1.0.0.0` will be the version of the NuGet package of Mapperly that you are using.
 
 Using these identification points, you can configure most tooling to apply a different set of rules to the generated source.
+
+## EditorConfig support
```
Comment (latonz) [inline docs/docs/configuration/generated-source.mdx:88]: Add a reference to the editorconfig section of the analyzer diagnostics docs page.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 2]
```diff
@@ -76,9 +83,53 @@ IncrementalValuesProvider<MapperNode> mappers
...
+                // Default to UTF-8 without BOM (most common for source files)
+                var encoding = _utf8NoBom;
+                if (!string.IsNullOrEmpty(mapper.Charset))
+                {
+                    encoding = mapper.Charset switch
+                    {
+                        "utf-8-bom" => _utf8WithBom, // UTF-8 with BOM
+                        "utf-16be" => Encoding.BigEndianUnicode,
+                        "utf-16le" => Encoding.Unicode,
+                        _ => _utf8NoBom, // utf-8 and others default to no BOM
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Nit: I don’t think these comments add much value, so they could probably be removed.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 3]
```diff
@@ -76,9 +83,53 @@ IncrementalValuesProvider<MapperNode> mappers
             mappers,
             static (spc, mapper) =>
             {
-                var mapperText = mapper.Body.GetText(Encoding.UTF8);
-                spc.AddSource(mapper.FileName, mapperText);
+                // Respect .editorconfig charset setting
+                // Default to UTF-8 without BOM (most common for source files)
+                var encoding = _utf8NoBom;
+                if (!string.IsNullOrEmpty(mapper.Charset))
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Is this if even needed? If the charset is null/empty/unknown the default arm of the switch is used anyway.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 2 above)
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Shouldn't we diagnostic if an unknown value is provided?

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 4]
```diff
@@ -76,9 +83,53 @@ IncrementalValuesProvider<MapperNode> mappers
...
         );
     }
+
+    private static string GetSourceText(Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax body, string? endOfLine)
+    {
+        var newLine = endOfLine switch
+        {
+            "lf" => "\n",
+            "cr" => "\r",
+            _ => null, // crlf, null, empty, or unknown - no replacement needed
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Shouldn't we diagnostic if an unknown value is provided?

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 5]
```diff
@@ -76,9 +83,53 @@ IncrementalValuesProvider<MapperNode> mappers
...
+                // For non-CRLF, use streaming replacement to avoid intermediate string allocation
+                var text = GetSourceText(mapper.Body, mapper.EndOfLine);
+
+                // Always use SourceText.From to ensure BOM is included when specified
+                spc.AddSource(mapper.FileName, SourceText.From(text, encoding));
             }
         );
     }
+
+    private static string GetSourceText(Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax body, string? endOfLine)
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Nit: is there a reason for using a FQN here instead of a using?

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 6]
```diff
@@ -76,9 +83,53 @@ IncrementalValuesProvider<MapperNode> mappers
...
+                    };
+                }
+
+                // Respect .editorconfig end_of_line setting
+                // The syntax tree uses CRLF by default (ElasticCarriageReturnLineFeed)
+                // For non-CRLF, use streaming replacement to avoid intermediate string allocation
+                var text = GetSourceText(mapper.Body, mapper.EndOfLine);
+
+                // Always use SourceText.From to ensure BOM is included when specified
+                spc.AddSource(mapper.FileName, SourceText.From(text, encoding));
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: This introduces a performance penalty even when the feature is not used. Could you measure the performance impact compared to main using `Riok.Mapperly.Benchmarks`, both with and without custom EOL or encoding settings?

File: src/Riok.Mapperly/Helpers/LineEndingTextWriter.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Riok.Mapperly.Helpers;
+
+/// <summary>
+/// A TextWriter that replaces CRLF (\r\n) with a target line ending on-the-fly.
+/// This avoids intermediate string allocations when converting line endings.
+/// </summary>
+internal sealed class LineEndingTextWriter : TextWriter
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/LineEndingTextWriter.cs]: Add unit tests to test this class.

File: src/Riok.Mapperly/Output/MapperNode.cs [hunk 8]
```diff
@@ -2,12 +2,15 @@
 
 namespace Riok.Mapperly.Output;
 
-public readonly record struct MapperNode(string FileName, CompilationUnitSyntax Body)
+public readonly record struct MapperNode(string FileName, CompilationUnitSyntax Body, string? EndOfLine = null, string? Charset = null)
```
Comment (latonz) [inline src/Riok.Mapperly/Output/MapperNode.cs]: It would probably make sense to introduce a separate `SourceTextConfig` record.

File: test/Riok.Mapperly.Tests/Generator/EndOfLineTest.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    [InlineData("utf-8-bom", true)]
+    public void ShouldRespectEditorConfigCharset(string? charsetSetting, bool expectBom)
+    {
+        var generatedText = GenerateSourceText(charset: charsetSetting);
+        var preamble = generatedText.Encoding?.GetPreamble() ?? [];
+
+        preamble.Length.ShouldBe(expectBom ? 3 : 0, $"charset={charsetSetting ?? "(default)"} should {(expectBom ? "" : "not ")}have BOM");
+    }
+
+    private static string GenerateSource(string? endOfLine = null, string? charset = null) =>
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/Generator/EndOfLineTest.cs]: This and the remaining code in this file should probably be integrated into or merged with `TestHelper`.

**Review by latonz — CHANGES_REQUESTED** [review]

File: docs/docs/configuration/generated-source.mdx (hunk 1 above)
Comment (YoelDruxman) [reply docs/docs/configuration/generated-source.mdx:88]: Done in fdf3973c6e93c41aca4bdd5d2d4d1d079773e6e0. Added cross-reference link to the analyzer diagnostics editorconfig section.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 2 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Done in a8993a4a9c4a76c81dba4f270812e56ef55b0f5d. Removed the inline comments as suggested.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 3 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Done in 75f17555b038f8865e1d01d06bd62afb12fc0ba8. Removed the unnecessary if check - the switch default arm handles null/empty/unknown values.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 2 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Done in 4efa273c4afb593639c8851d72d8bfc11550d8a2. Added RMG095 diagnostic (Info level) for unknown charset values.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 4 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Done in 4efa273c4afb593639c8851d72d8bfc11550d8a2. Added RMG096 diagnostic (Info level) for unknown end_of_line values.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 5 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Done in 58af35e62b15a5ccaee3351a513b7f2239df24b3. Replaced FQN with using directive for CompilationUnitSyntax.

File: src/Riok.Mapperly/Helpers/LineEndingTextWriter.cs (hunk 7 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/LineEndingTextWriter.cs]: Done in d8fe736dbc4410e73f13f41d63bc760a7891e041. Added comprehensive unit tests for LineEndingTextWriter covering CRLF conversion, edge cases, and buffer boundary handling. Made the class public to allow direct testing, following the pattern of other tested helper classes.

File: src/Riok.Mapperly/Output/MapperNode.cs (hunk 8 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Output/MapperNode.cs]: Done in a899c38ff4c4bdac1bbe55f43f3411cd23881a86. Introduced SourceTextConfig record to encapsulate EndOfLine and Charset settings.

File: test/Riok.Mapperly.Tests/Generator/EndOfLineTest.cs (hunk 9 above)
Comment (YoelDruxman) [reply test/Riok.Mapperly.Tests/Generator/EndOfLineTest.cs]: Done in 40b791d44b09b6d8b56a3c484b51f73f3956381c. Integrated editorconfig test infrastructure with TestHelper. Added SourceTextConfig parameter to TestHelperOptions and moved TestAnalyzerConfigOptionsProvider to a shared location.

File: src/Riok.Mapperly/AnalyzerReleases.Unshipped.md [hunk 10]
```diff
@@ -1,2 +1,9 @@
 ; Unshipped analyzer release
 ; https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md
+
+### New Rules
+
+Rule ID | Category | Severity | Notes
```
Comment (latonz) [inline src/Riok.Mapperly/AnalyzerReleases.Unshipped.md]: We dont use the unshipped file, we track added/removed diagnostics via git. Please move it to the shipped file.

File: src/Riok.Mapperly/MapperGenerator.cs [hunk 11]
```diff
@@ -110,18 +116,53 @@ CancellationToken cancellationToken
...
+            // Users should configure settings in [*.cs] or [*] sections
+            var generatedFileName = compilationContext.FileNameBuilder.Build(descriptor);
+            var configOptions = configOptionsProvider.GetOptions(mapperDeclaration.Syntax.SyntaxTree);
+            configOptions.TryGetValue("end_of_line", out var endOfLine);
+            configOptions.TryGetValue("charset", out var charset);
+            var sourceTextConfig = new SourceTextConfig(endOfLine, charset);
+
+            // Validate editorconfig settings and add diagnostics for unknown values
+            var location = mapperDeclaration.Syntax.GetLocation();
+            var allDiagnostics = diagnostics.ToList();
```
Comment (latonz) [inline src/Riok.Mapperly/MapperGenerator.cs]: Why the ToList? Couldn't you report it directly via `DiagnosticCollection`?

File: src/Riok.Mapperly/MapperGenerator.cs [hunk 12]
```diff
@@ -110,18 +116,53 @@ CancellationToken cancellationToken
...
+            var mapper = new MapperNode(generatedFileName, SourceEmitter.Build(descriptor, cancellationToken), sourceTextConfig);
+            return new MapperAndDiagnostics(mapper, allDiagnostics.ToImmutableEquatableArray());
         }
         catch (OperationCanceledException)
         {
             return null;
         }
     }
 
+    private static readonly HashSet<string> _validCharsets = new(StringComparer.OrdinalIgnoreCase)
```
Comment (latonz) [inline src/Riok.Mapperly/MapperGenerator.cs]: This codebase keeps the fields on the top of the class.

File: src/Riok.Mapperly/MapperGenerator.cs [hunk 13]
```diff
@@ -110,18 +116,53 @@ CancellationToken cancellationToken
...
+    {
+        "utf-8",
+        "utf-8-bom",
+        "utf-16be",
+        "utf-16le",
+    };
+
+    private static readonly HashSet<string> _validEndOfLineValues = new(StringComparer.OrdinalIgnoreCase) { "lf", "crlf", "cr" };
```
Comment (latonz) [inline src/Riok.Mapperly/MapperGenerator.cs]: Now we have these constants at multiple locations... Wouldn't it make sense to map to the actual value here now? Move the switch here, the default arm reports the diagnostic, `SourceTextConfig.EndOfLine` is `\n`/`\r\n`/... and `SourceTextConfig.Charset` would get `SourceTextConfig.Encoding`.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 14]
```diff
@@ -62,6 +64,12 @@ IncrementalValuesProvider<ImmutableEquatableArray<Diagnostic>> diagnostics
         );
     }
 
+    // UTF-8 encoding without BOM (default for 'charset = utf-8' in .editorconfig)
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: This codebase keeps fields at the top of a class.

File: test/Riok.Mapperly.Tests/TestHelperOptions.cs [hunk 15]
```diff
@@ -9,13 +9,17 @@ namespace Riok.Mapperly.Tests;
...
 public record TestHelperOptions(
     NullableContextOptions NullableOption = NullableContextOptions.Enable,
     LanguageVersion LanguageVersion = LanguageVersion.Default,
     IReadOnlySet<DiagnosticSeverity>? AllowedDiagnosticSeverities = null,
     IReadOnlySet<DiagnosticDescriptor>? IgnoredDiagnostics = null,
     string AssemblyName = "Tests",
-    string GeneratedTreeFileName = $"{TestSourceBuilderOptions.DefaultMapperClassName}.g.cs"
+    string GeneratedTreeFileName = $"{TestSourceBuilderOptions.DefaultMapperClassName}.g.cs",
+    string? EditorConfigEndOfLine = null,
+    string? EditorConfigCharset = null
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/TestHelperOptions.cs:22]: Wouldn't it make more sense using the `SourceTextConfig` here?

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 16]
```diff
@@ -76,9 +84,49 @@ IncrementalValuesProvider<MapperNode> mappers
...
+                    "utf-8-bom" => _utf8WithBom,
+                    "utf-16be" => Encoding.BigEndianUnicode,
+                    "utf-16le" => Encoding.Unicode,
+                    _ => _utf8NoBom,
+                };
+
+                // Respect .editorconfig end_of_line setting
+                // The syntax tree uses CRLF by default (ElasticCarriageReturnLineFeed)
+                // For non-CRLF, use streaming replacement to avoid intermediate string allocation
+                var text = GetSourceText(mapper.Body, mapper.SourceTextConfig.EndOfLine);
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: IMO we should add a fast path using the `GetText` directly if no option or only the encoding is set.

File: src/Riok.Mapperly/AnalyzerReleases.Unshipped.md (hunk 10 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/AnalyzerReleases.Unshipped.md]: Done in 902cfb51792827fa08bc85c23004890387b31cd2

File: src/Riok.Mapperly/MapperGenerator.cs (hunk 12 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/MapperGenerator.cs]: Done in 51ed304051f45ad6845a8bd9aee8914832803823

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 14 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Done in e4c47469cc0024532ec34bc9a986205f2c5414b1

File: src/Riok.Mapperly/MapperGenerator.cs (hunk 11 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/MapperGenerator.cs]: Done in 11a2002

File: src/Riok.Mapperly/MapperGenerator.cs (hunk 13 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/MapperGenerator.cs]: Done in 11a2002

File: test/Riok.Mapperly.Tests/TestHelperOptions.cs (hunk 15 above)
Comment (YoelDruxman) [reply test/Riok.Mapperly.Tests/TestHelperOptions.cs:22]: The test options use raw editorconfig strings (`"lf"`, `"utf-8"`, etc.) while `SourceTextConfig` holds actual parsed values (`"\n"`, `Encoding.UTF8`, etc.).

The current approach feeds raw strings to `TestAnalyzerConfigOptionsProvider` which simulates the editorconfig file, then the generator parses them into `SourceTextConfig`. This tests the full editorconfig parsing path.

If we use `SourceTextConfig` directly, we'd either need to convert actual values back to raw strings (seems backwards), or bypass the editorconfig parsing layer entirely and inject `SourceTextConfig` directly into the generator (significant infrastructure change).

Could you clarify what you had in mind?

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 16 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Great idea! Done in 11a2002 - fast path now uses `body.GetText(config.Encoding)` directly when no line ending conversion is needed (`EndOfLine is null or "\r\n"`).

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 6 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: I updated the code to have the fast-track as requested, when measuring performance, there did not seem to be any impact in any cases (although I am not 100% confident).

**Comment by YoelDruxman** [comment]
@latonz Thank you for the review! I think I addressed all your comments. Please let me know if there is more to do. As I replied above, I was not able to detect any meaningful performance difference.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 6 above)
Comment (latonz) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Could you post the benchmark dotnet results?

File: test/Riok.Mapperly.Tests/TestHelperOptions.cs (hunk 15 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/TestHelperOptions.cs:22]: Makes sense, we will keep it as is.

File: src/Riok.Mapperly/Helpers/LineEndingTextWriter.cs [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+        _sb.Append(value);
+    }
+
+    public override void Write(string? value)
+    {
+        if (value == null)
+            return;
+
+        foreach (var c in value)
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/LineEndingTextWriter.cs:59]: Using foreach on a string allocates an enumerator. Consider using a for loop with indexing to avoid the allocation, especially since this method may be called frequently during code generation.

File: test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    public override AnalyzerConfigOptions GetOptions(SyntaxTree tree) => _fileOptions;
+
+    public override AnalyzerConfigOptions GetOptions(AdditionalText textFile) => _fileOptions;
+}
+
+/// <summary>
+/// A test implementation of <see cref="AnalyzerConfigOptions"/> that provides
+/// specific editorconfig values for testing.
+/// </summary>
+internal class TestAnalyzerConfigOptions : AnalyzerConfigOptions
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs:32]: This codebase follows a file per class approach, please move it into its own file.

File: test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+{
+    public static readonly TestAnalyzerConfigOptions Empty = new(null, null);
+
+    private readonly ImmutableDictionary<string, string> _options;
+
+    public TestAnalyzerConfigOptions(string? endOfLine, string? charset)
+    {
+        _options = new Dictionary<string, string?> { ["end_of_line"] = endOfLine, ["charset"] = charset }
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs:40]: The dictionary initialization could be improved by using an immutable dictionary builder and avoiding the intermediate nullable dictionary.

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs [hunk 20]
```diff
@@ -76,9 +78,25 @@ IncrementalValuesProvider<MapperNode> mappers
...
+            return body.GetText(config.Encoding);
+        }
+
+        var sb = new StringBuilder();
+        using (var writer = new LineEndingTextWriter(sb, config.EndOfLine))
+        {
+            body.WriteTo(writer);
+        }
+
+        return SourceText.From(sb.ToString(), config.Encoding);
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs:100]: To improve performance, the initial capacity of the `StringBuilder` should be set.

Performance could probably be further improved by using a pipelines implementation using the `SourceText.From` overload that accepts a `TextReader`, which allocates less memory for large texts (I think over 40 KB) or a custom `SourceText` implementation. But I think these approaches are not worth it for now.

File: src/Riok.Mapperly/MapperGenerator.cs [hunk 21]
```diff
@@ -122,6 +139,44 @@ CancellationToken cancellationToken
...
     }
 
+    private static SourceTextConfig BuildSourceTextConfig(AnalyzerConfigOptions configOptions, DiagnosticCollection diagnostics)
+    {
+        configOptions.TryGetValue("end_of_line", out var endOfLineValue);
+        configOptions.TryGetValue("charset", out var charsetValue);
+
+        var endOfLine = endOfLineValue?.ToLowerInvariant() switch
+        {
+            null or "" => null,
```
Comment (latonz) [inline src/Riok.Mapperly/MapperGenerator.cs:149]: The handling on the reading side could be simplified by eliminating the null case by using `\r\n` here (combining it with the `crlf` case). `ReportUnknownEndOfLine` could also just return `\r\n`

File: src/Riok.Mapperly/MapperGenerator.cs [hunk 22]
```diff
@@ -122,6 +139,44 @@ CancellationToken cancellationToken
...
+            _ => ReportUnknownEndOfLine(endOfLineValue!, diagnostics),
+        };
+
+        var encoding = charsetValue?.ToLowerInvariant() switch
+        {
+            null or "" or "utf-8" => _utf8NoBom,
+            "utf-8-bom" => _utf8WithBom,
+            "utf-16be" => Encoding.BigEndianUnicode,
+            "utf-16le" => Encoding.Unicode,
+            _ => ReportUnknownCharset(charsetValue!, diagnostics),
```
Comment (latonz) [inline src/Riok.Mapperly/MapperGenerator.cs:162]: The handling on the reading side could be simplified by eliminating the null case by returning `_utf8NoBom` from `ReportUnknownCharset`.

File: test/Riok.Mapperly.Tests/Generator/EndOfLineTest.cs [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+    [InlineData("crlf", "\r\n")]
+    [InlineData("cr", "\r")]
+    public void ShouldRespectEditorConfigEndOfLine(string? endOfLineSetting, string expectedLineEnding)
+    {
+        var generatedSource = GenerateSource(endOfLine: endOfLineSetting);
+
+        generatedSource.ShouldContain(expectedLineEnding);
+
+        // After removing expected line endings, no line ending chars should remain
+        var normalized = generatedSource.Replace(expectedLineEnding, "");
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/Generator/EndOfLineTest.cs:20]: Nit: This codebase uses `string.Empty` for empty string literals.

File: test/Riok.Mapperly.Tests/TestHelper.cs [hunk 24]
```diff
@@ -26,6 +27,16 @@ public static Task<VerifyResult> VerifyGenerator(string source, TestHelperOption
...
     }
 
+    /// <summary>
+    /// Generates mapper source and returns the first generated SourceText.
+    /// Useful for testing encoding and line ending settings.
+    /// </summary>
+    public static SourceText GenerateSourceText(string source, TestHelperOptions? options = null)
+    {
+        var driver = Generate(source, options);
+        return driver.GetRunResult().GeneratedTrees.First().GetText();
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/TestHelper.cs:37]: Nit: wouldn't it be saver to use `Single` instead of `First`?

File: test/Riok.Mapperly.Tests/TestHelper.cs [hunk 25]
```diff
@@ -100,7 +112,13 @@ private static GeneratorDriver Generate(
 
         var generator = new MapperGenerator();
 
-        GeneratorDriver driver = CSharpGeneratorDriver.Create(generator);
+        // Create options provider for editorconfig settings if specified
+        var optionsProvider =
+            options.EditorConfigEndOfLine != null || options.EditorConfigCharset != null
+                ? new TestAnalyzerConfigOptionsProvider(options.EditorConfigEndOfLine, options.EditorConfigCharset)
+                : null;
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/TestHelper.cs:119]: Are the null checks really required? Wouldn't always using `new TestAnalyzerConfigOptionsProvider(options.EditorConfigEndOfLine, options.EditorConfigCharset)` lead to the same (`TryGetValue` returning `false`)?

File: src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs (hunk 6 above)
Comment (YoelDruxman) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Here are the benchmark results:

```
BenchmarkDotNet v0.15.8, Windows 11
AMD Ryzen 9 7900X 4.70GHz, 1 CPU, 24 logical and 12 physical cores
.NET SDK 10.0.102
  [Host] : .NET 10.0.2, X64 RyuJIT x86-64-v4

| Method       | Mean      | Error     | StdDev    | Gen0      | Gen1     | Allocated   |
|------------- |----------:|----------:|----------:|----------:|---------:|------------:|
| Compile      |  1.340 ms | 0.0264 ms | 0.0475 ms |   46.8750 |  15.6250 |   818.41 KB |
| LargeCompile | 38.092 ms | 0.7134 ms | 0.6673 ms | 1384.6154 | 692.3077 | 23124.07 KB |
```

Comment (latonz) [reply src/Riok.Mapperly/Helpers/IncrementalValuesProviderExtensions.cs]: Thanks for this. The impact here looks negligible.

For feature PRs that could affect performance, please run the benchmarks and include the results and the diff versus the main branch. That makes it much easier to evaluate the change / performance impact.

File: test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs (hunk 19 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs:40]: I implemented similar options recently in [#2142](https://github.com/riok/mapperly/pull/2142/changes#diff-26b7fa9448971ae0416595aafb2b471173b787e7f477ada8b0ce05e7a2e2070bR18). Once you rebase, could you please remove your added classes and use the existing dictionary `TestHelperOptions.AnalyzerConfigOptions` instead.

File: test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs (hunk 18 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/TestAnalyzerConfigOptionsProvider.cs:32]: See https://github.com/riok/mapperly/pull/2114#discussion_r2769994279

File: test/Riok.Mapperly.Tests/TestHelper.cs (hunk 25 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/TestHelper.cs:119]: See https://github.com/riok/mapperly/pull/2114#discussion_r2769994279

**Comment by YoelDruxman** [comment]
Hello. I ended up finding that we could just emit all the generated files in CRLF, even though the rest is in LF, and update editor config to expect that. I no longer need this feature, even though with it my whole codebase could be in LF. It's not a priority for me right now. @latonz
